active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,251 @@
+
+
+
+# Examples
+
+## IProvideValueTarget
+
+The following example shows a custom markup extension whose provided value changes based on the specific property being targeted by the custom markup extension. In this case, the developer wants something that will automatically use an AcrylicBrush for the [Control.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Control.Background) or [Border.Background](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Border.Background) properties, but uses a [SolidColorBrush](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.SolidColorBrush) for all other properties.
+
+```cs
+public class BrushSelectorExtension : MarkupExtension
+{
+    public Color Color { get; set; }
+
+    protected override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        Brush brushToReturn = new SolidColorBrush() { Color = Color }; ;
+
+        var provideValueTarget = serviceProvider.GetService(typeof(IProvideValueTarget))
+            as IProvideValueTarget;
+        if (provideValueTarget.TargetProperty is ProvideValueTargetProperty targetProperty)
+        {
+            if (targetProperty.Name == "Background"
+                && (targetProperty.DeclaringType == typeof(Control) 
+                    || targetProperty.DeclaringType == typeof(Border)))
+            {
+                brushToReturn = new AcrylicBrush()
+                    { TintColor = Color, TintOpacity = 0.75 };
+            }
+        }
+
+        return brushToReturn;
+    }
+}
+```
+
+```xml
+<StackPanel>
+    <Button Foreground="{local:BrushSelector Color=Blue}" 
+            Background="{local:BrushSelector Color=Gold}">
+        Go bears!
+    </Button>
+    <Rectangle x:Name="SolidColor" Fill="{local:BrushSelector Color=Green}" />
+</StackPanel>
+```
+
+
+## IRootObjectProvider
+
+> Some more realistic example?
+
+For example, with this markup extension:
+
+```cs
+public class TestMarkupExtension : MarkupExtension
+{
+    public override object ProvideValue(IXamlServiceProvider serviceProvider)
+    {
+        var target = serviceProvider.GetService(typeof(IRootObjectProvider)) as IRootObjectProvider;
+        return target.RootObject.ToString();
+    }
+}
+```
+
+the TextBlock in this markup will display “App1.MainPage”:
+
+```xml
+<Page x:Class="App52.MainPage"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:App52"
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}" >
+
+    <Grid>
+        <TextBlock Text="{local:TestMarkupExtension}" />
+    </Grid>
+</Page>
+```
+
+## IUriContext 
+
+> Todo
+
+# API Notes/Remarks
+
+## IXamlServiceProvider interface
+Gets the service object of the specified type.
+
+There are two versions of the virtual ProvideValue method, you should only override one.

```

Comment:
Can a warning be raised if a developer overrides both versions?

Reply:
I don't see a way, all the base class can do is call it. This is the same "fun" we had with DataTemplateSelector.SelectTemplateCore.